### PR TITLE
Add vault.state testing on takeLoan + some refactors

### PR DESCRIPTION
**Pull request by canonbrother**
#### What kind of PR is this?:

/kind refactor
/kind feature

#### What this PR does / why we need it:
Added vault.state test on takeLoan
Added no live fixed price test
Added test exceed collateral ratio by taking multi loans 
Added liquidated by interest
Added vault.batches splitting value after liquidation on taking multiple loans
Refactored takeLoan rpc and txn builder

### CODE REVIEW

**Comment by jingyi2811** [comment]
Add 1 more test => Take loan without collaterals. You should get the error below.

Screenshot 2021-11-05 at 4 40 10 PM

File: packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts [hunk 1]
```diff
@@ -3,279 +3,636 @@ import { GenesisKeys } from '@defichain/testcontainers'
...
+  expect(bobVault.interestValue).toStrictEqual(new BigNumber(0))
+  expect(bobVault.collateralRatio).toStrictEqual(-1) // empty loan
+  expect(bobVault.informativeRatio).toStrictEqual(new BigNumber(-1)) // empty loan
+}
+
+describe('takeLoan success', () => {
+  beforeEach(async () => {
+    await tGroup.start()
+    await alice.container.waitForWalletCoinbaseMaturity()
+    await setup()
```
Comment (ivan-zynesis) [inline packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts:207]: Can replace with a simple single vault setup here. The complete `setup` for each fail cases are heavy and repeated here without being used.

File: packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts [hunk 2]
```diff
@@ -3,279 +3,636 @@ import { GenesisKeys } from '@defichain/testcontainers'
...
+  })
 
-    it('should not takeLoan on invalid token', async () => {
-      const promise = tGroup.get(0).rpc.loan.takeLoan({
-        vaultId: vaultId,
-        amounts: '1@INVALID'
-      })
-      await expect(promise).rejects.toThrow(RpcApiError)
-      await expect(promise).rejects.toThrow('Invalid Defi token: INVALID')
+  it('should not takeLoan on mayLiquidate vault', async () => {
```
Comment (ivan-zynesis) [inline packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts:469]: This test is perhaps less meaningful, `takeLoan` will still go through as long as resulted total amount within ratio.

Suggest test to be added
`should not takeLoan if total loan exceeded min collateral ratio` with taking multiple loan token but each within ratio.

Comment (canonbrother) [reply packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts:612]: the mayLiquidate test can be used to assert the error message
and agree on testing the `liquidated by take multi loans`, it has extra logic... added

File: packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts [hunk 3]
```diff
@@ -3,279 +3,803 @@ import { GenesisKeys } from '@defichain/testcontainers'
...
+    expect(bobVault.loanSchemeId).toStrictEqual('scheme')
+    expect(bobVault.ownerAddress).toStrictEqual(bobVaultAddr)
+    expect(bobVault.state).toStrictEqual('active')
+    expect(bobVault.collateralAmounts).toStrictEqual(['10000.00000000@DFI', '1.00000000@BTC'])
+    expect(bobVault.collateralValue).toStrictEqual(new BigNumber(15000))
+    expect(bobVault.loanAmounts).toStrictEqual([])
+    expect(bobVault.loanValue).toStrictEqual(new BigNumber(0))
+    expect(bobVault.interestAmounts).toStrictEqual([])
+    expect(bobVault.interestValue).toStrictEqual(new BigNumber(0))
+    expect(bobVault.collateralRatio).toStrictEqual(-1) // empty loan
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts:140]: This should be `new BigNumber(-1)`. You may need to change the return type for `collateralRatio` in `getVault` RPC

Comment (canonbrother) [reply packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts:140]: hmm.. have `collateralRatio` got decimal??

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts:140]: LOL. Okay, then make sense to keep as number.

**Review by jingyi2811 — APPROVED** [review]

**Review by surangap — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**